fl4health/utils/sampler.py

```diff
@@ -25,9 +26,10 @@ def __init__(self, unique_labels: List[Any]) -> None:
 
     def select_by_indices(self, dataset: D, selected_indices: torch.Tensor) -> D:
         if isinstance(dataset, BaseDataset):
-            dataset.targets = dataset.targets[selected_indices]
-            dataset.data = dataset.data[selected_indices]
-            return dataset
+            modified_dataset = copy.deepcopy(dataset)
+            modified_dataset.targets = dataset.targets[selected_indices]
+            modified_dataset.data = dataset.data[selected_indices]
+            return modified_dataset
```

Comment:
This is fixing a "bug" in the sense that it was returning a dataset but also modifying the dataset passed in-place, which is not what a user might expect. We should either modify the dataset in place and return None or create a new dataset and leave the one that was passed in alone. I went with the former here. It shouldn't have any affect on our examples/tests etc., since in all existing cases we're just sampling once at most. However, the fact that we're not modifying in-place is leveraged for reuse in the tests below.

Reply:
That is a nice catch, though I think my experiments should not be affected by it. Correct me if I'm wrong.

Reply:
Yeah I don't think it should affect your experiments. It's specifically problematic if you so sampling on the dataset multiple times, which I don't think you're doing in your experiments?

Reply:
I agree